py/gc.c

```diff
@@ -290,14 +271,56 @@ STATIC bool gc_try_add_heap(size_t failed_alloc) {
         total_heap += ALLOC_TABLE_GAP_BYTE + sizeof(mp_state_mem_area_t);
     }
 
-    DEBUG_printf("total_heap " UINT_FMT " bytes\n", total_heap);
+    // Deciding how much to allocate for the new "split" heap is tricky:
+    //
+    // - Grow by too small amounts, leads to heap fragmentation issues.
+    //
+    // - Grow by too large amounts, may lead to system heap running out of
+    //   space.
+    //
+
+    // Start by choosing the current total Python heap size for the new heap.
+    // With no other constraints, the total Python heap size would double each
+    // time: i.e 64KB -> 128KB -> 256KB -> 512KB -> 1MB, etc. This avoids
+    // fragmentation where possible.
+
+    size_t new_heap_size = total_heap;
+
+    // If this size is >75% of the total free RAM (across all free blocks of
+    // system heap), reduce it to 75% of the total free RAM. This is to avoid
+    // greedily claiming all of the system RAM for Python heap and running
+    // into allocation failures elsewhere in the system.
+
+    size_t total_free = gc_get_total_free();
+    size_t grow_limit = total_free * 75 / 100;
```

Comment:
Could this grow limit be implemented to be configurable? People who want to use all the memory possible aren't limited in those instances, and advanced users can configure this to whatever value they like.

Reply:
Thanks @bobveringa, it's an idea worth considering for sure.

To be clear on a small difference, this isn't a hard limit - the Python heap will still grow to use all of the available RAM if the alternative is a `MemoryError`.

This limit is to limit "greedily" adding more memory than is needed to the Python heap, it eases back to avoid prematurely taking all of the system heap for Python.